### PR TITLE
Enable iterative viscosity dampening

### PR DESCRIPTION
**Pull request by anne-glerum**
Hi John, I've updated your iterative viscosity damping PR and am testing it now. For very small problems is does not seem to do much yet, so I'm running it on the Buiter et al. 2016 brittle wedge benchmark now. I'll let you know!

I'm creating this PR so that we can discuss it, but I might have to push directly to your PR on the ASPECT repo as I've rebased to main. I think that is also why it looks like I changed 1788 files :/

### Before your first pull request:

* [ ] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [ ] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: source/material_model/visco_plastic.cc [hunk 1]
```diff
@@ -284,24 +284,27 @@ namespace aspect
...
+                PrescribedFieldOutputs<dim> *prescribed_field_out = out.template get_additional_output<PrescribedFieldOutputs<dim>>();
+
+                if (in.current_cell.state() == IteratorState::valid && prescribed_field_out != NULL)
+                  {
+                    const int field_index = this->introspection().compositional_index_for_name("viscosity_field");
+                    const double old_viscosity = in.composition[i][field_index];
+
+                    // Only dampen the viscosity after the first nonlinear iteration.
+                    if (this->simulator_is_past_initialization() == true &&
+                        this->get_timestep_number() > 0 &&
```
Comment (naliboff) [inline source/material_model/visco_plastic.cc]: I originally also asserted that this should only happen after the first time step, but then switched to doing this on all time steps. What is your reasoning for not doing the dampening on the first time step (I don't recall what my logic was)?

**Review by naliboff — COMMENTED** [review]
@anne-glerum - Thanks for these changes, which all make sense at first glance. I rebased my branch, and then pushed to the PR so that only the original commits and your commits show. 

Shall I go ahead and merge this PR, and then we can continue jointly testing?

File: source/material_model/visco_plastic.cc (hunk 1 above)
Comment (anne-glerum) [reply source/material_model/visco_plastic.cc]: I figured that at t0 the initial guesses for velocity/strain rate and pressure are so bad that the viscosity is also far off and changes rapidly with each subsequent nonlinear iteration, and therefore averaging those viscosities would not make the convergence better. But I don't know if that's true.

On a sidenote, we're aiming at situations where the residual oscillates, right, which usually only happens after a certain number of nonlinear iterations within a timestep. Maybe we can set `this->get_nonlinear_iteration() > 5` (or some other value larger than 0), but do apply the dampening to each time step?

**Comment by anne-glerum** [comment]
Hi @naliboff, thank you for having a look and for rebasing. Do you know why your commits show up as commits in my pull request?

File: source/material_model/visco_plastic.cc (hunk 1 above)
Comment (naliboff) [reply source/material_model/visco_plastic.cc]: > I figured that at t0 the initial guesses for velocity/strain rate and pressure are so bad that the viscosity is also far off and changes rapidly with each subsequent nonlinear iteration, and therefore averaging those viscosities would make the convergence better. But I don't know if that's true.

That makes sense. If needed, we can always add a parameter that allows switching the dampening on/off on the first time step.

> Maybe we can set this->get_nonlinear_iteration() > 5 (or some other value larger than 0), but do apply the dampening to each time step?

Yep, that sounds like a good strategy.

**Comment by naliboff** [comment]
> Hi @naliboff, thank you for having a look and for rebasing. Do you know why your commits show up as commits in my pull request?

I'm not sure at all and was wondering the same thing. I think this was the case even before I rebased, but it may have also happened as a result of it. 

I propose the following:
1. You close the PR.
2. We both rebase to main
4. You reopen the PR and then we can see if the issue persists.

Does that sound reasonable?

File: source/material_model/visco_plastic.cc (hunk 1 above)
Comment (anne-glerum) [reply source/material_model/visco_plastic.cc]: Alright, then after we've rebased, I'll switch to applying averaging after the fifth iteration for each timestep, including the first.

**Pull request closed without merging**